src/gui/DownloadIcon.cpp

```diff
@@ -0,0 +1,217 @@
+/*
+ *  Copyright (C) 2019 KeePassXC Team <team@keepassxc.org>
+ *
+ *  This program is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 2 or (at your option)
+ *  version 3 of the License.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "DownloadIcon.h"
+#include "core/Config.h"
+
+#ifdef WITH_XC_NETWORKING
+#include <QHostInfo>
+#include <QNetworkAccessManager>
+#include <QtNetwork>
+#endif
+
+DownloadIcon::DownloadIcon(QObject* parent)
+    : QObject(parent)
+#ifdef WITH_XC_NETWORKING
+    , m_netMgr(new QNetworkAccessManager(this))
+    , m_reply(nullptr)
+#endif
+{
+}
+
+DownloadIcon::~DownloadIcon()
+{
+}
+
+#ifdef WITH_XC_NETWORKING
+namespace
+{
+    // Try to get the 2nd level domain of the host part of a QUrl. For example,
+    // "foo.bar.example.com" would become "example.com", and "foo.bar.example.co.uk"
+    // would become "example.co.uk".
+    QString getSecondLevelDomain(const QUrl& url)
+    {
+        QString fqdn = url.host();
+        fqdn.truncate(fqdn.length() - url.topLevelDomain().length());
+        QStringList parts = fqdn.split('.');
+        QString newdom = parts.takeLast() + url.topLevelDomain();
+        return newdom;
+    }
+
+    QUrl convertVariantToUrl(const QVariant& var)
+    {
+        QUrl url;
+        if (var.canConvert<QUrl>())
+            url = var.toUrl();
+        return url;
+    }
+
+    QUrl getRedirectTarget(QNetworkReply* reply)
+    {
+        QVariant var = reply->attribute(QNetworkRequest::RedirectionTargetAttribute);
+        QUrl url = convertVariantToUrl(var);
+        return url;
+    }
+} // namespace
+#endif
+
+void DownloadIcon::downloadFavicon(Entry* entry)
+{
+#ifdef WITH_XC_NETWORKING
+    m_entry = entry;
+    m_url = entry->url();
+    m_redirects = 0;
+    m_urlsToTry.clear();
+
+    if (!m_url.scheme().startsWith("http")) {
```

Comment:
If I use a non-standard scheme, like chrome://, this will give me the favicon from duckduckgo.com. We should only add the scheme if it doesn't exist already.